From 02405ebf6a05d53658b3c8b02ae60e69e16297cf Mon Sep 17 00:00:00 2001
From: Mike Gerdts <mike.gerdts@joyent.com>
Date: Wed, 7 Nov 2018 17:46:44 +0000
Subject: [PATCH] OS-7347 default bhyve quota not used while creating a bhyve
 instance Reviewed by: Orlando Vazquez <orlando@joyent.com> Reviewed by: Jorge
 Schrauwen <sjorge@blackdot.be> Reviewed by: Dylan Yep <dylan.yep@joyent.com>
 Approved by: Dylan Yep <dylan.yep@joyent.com>

---
 src/vm/node_modules/VM.js  | 14 +++++++++----
 src/vm/tests/test-quota.js | 41 +++++++++++++++++++++++++++++++++++---
 2 files changed, 48 insertions(+), 7 deletions(-)

diff --git a/src/vm/node_modules/VM.js b/src/vm/node_modules/VM.js
index 2403d984..c18e0efb 100644
--- a/src/vm/node_modules/VM.js
+++ b/src/vm/node_modules/VM.js
@@ -137,11 +137,12 @@ exports.KVM_MEM_OVERHEAD = KVM_MEM_OVERHEAD;
  *
  */
 
+var DEFAULT_QUOTA = 10;                     /* GiB */
+var DEFAULT_BHYVE_QUOTA = 1;                /* GiB */
 var DEFAULT_MAX_MSG_IDS = 4096;
 var DEFAULT_MAX_SEM_IDS = 4096;
 var DEFAULT_MAX_SHM_IDS = 4096;
 var DEFAULT_MDATA_TIMEOUT = 300;
-var DEFAULT_BHYVE_QUOTA = (1024 * 1024 * 1024);
 var DISABLED = 0;
 var MAX_HOSTVOL_FILE_BYTES = (10 * 1024 * 1024);
 var MAX_SNAPNAME_LENGTH = 64;
@@ -1468,8 +1469,9 @@ function setQuotaBhyve(opts, callback) {
     var volrefres;
 
 
-    var maxQuota = Math.max(DEFAULT_BHYVE_QUOTA,
-        opts.quota * 1024 * 1024 * 1024);
+    // maxQuota is in GiB
+    var maxQuota =
+        Math.max(DEFAULT_BHYVE_QUOTA, opts.quota) * 1024 * 1024 * 1024;
     // flexible_disk_size is in MiB to match disk.*.size units.
     var flexsize = opts.flexible_disk_size
         ? opts.flexible_disk_size * 1024 * 1024 :  0;
@@ -4647,7 +4649,11 @@ function applyZoneDefaults(payload, log)
     }
 
     if (!payload.hasOwnProperty('quota')) {
-        payload.quota = 10; // in GiB
+        if (payload.brand === 'bhyve') {
+            payload.quota = DEFAULT_BHYVE_QUOTA;
+        } else {
+            payload.quota = DEFAULT_QUOTA;
+        }
     }
 
     if (!payload.hasOwnProperty('billing_id')) {
diff --git a/src/vm/tests/test-quota.js b/src/vm/tests/test-quota.js
index c38dd143..b1fbbb15 100644
--- a/src/vm/tests/test-quota.js
+++ b/src/vm/tests/test-quota.js
@@ -114,6 +114,43 @@ var test_cases = [
             flexible_disk_size: 12 * 1024, kernel_version: '3.13.0',
             image_uuid: vmtest.CURRENT_UBUNTU_LX_IMAGE_UUID},
         expected: {quota: 10, flexible_disk_size: undefined}
+    },
+
+    /*
+     * When quota is not specified, bhyve defaults to 1 GiB, others default to
+     * 10 GiB.
+     */
+    {
+        payload: {
+            brand: 'bhyve',
+            disks: [ {boot: true, model: 'virtio',
+                image_uuid: vmtest.CURRENT_BHYVE_CENTOS_UUID} ]
+        },
+        expected: {quota: 1}
+    },
+    {
+        payload: {
+            brand: 'kvm',
+            disks: [ {boot: true, model: 'virtio',
+                image_uuid: vmtest.CURRENT_BHYVE_CENTOS_UUID} ]
+        },
+        expected: {quota: 10}
+    },
+    {
+        payload: {brand: 'joyent', image_uuid: smartos_image_uuid},
+        expected: {quota: 10}
+    },
+    {
+        payload: {brand: 'joyent-minimal', image_uuid: smartos_image_uuid},
+        expected: {quota: 10}
+    },
+    {
+        payload: {
+            brand: 'lx',
+            kernel_version: '3.13.0',
+            image_uuid: vmtest.CURRENT_UBUNTU_LX_IMAGE_UUID
+        },
+        expected: {quota: 10}
     }
 ];
 
@@ -319,9 +356,7 @@ function checkProps(opts, callback) {
 
         Object.keys(expected).forEach(function (prop) {
             if (skipProps.indexOf(prop) === -1) {
-                t.ok(expected[prop] === o[prop],
-                    'correct ' + prop + ' [' + o[prop] + ','
-                    + expected[prop] + ']');
+                t.equal(expected[prop], o[prop], 'correct ' + prop);
             }
         });
         callback();
-- 
2.21.0

